### PR TITLE
aya: Support multiple maps in map sections

### PR DESCRIPTION
**Pull request by dave-tucker**
This commit uses the symbol table to discover all maps inside an ELF
section. We then divide the data in to equal sized chunks and parse them
using parse_map_def

Signed-off-by: Dave Tucker <dave@dtucker.co.uk>

### CODE REVIEW

**Comment by alessandrod** [comment]
I'll review this soon but first - have you tried it with stripped binaries?

**Comment by dave-tucker** [comment]
I haven't, but I can do

File: aya/src/obj/mod.rs [hunk 1]
```diff
@@ -600,6 +642,15 @@ pub enum ParseError {
 
     #[error("map for section with index {index} not found")]
     MapNotFound { index: usize },
+
+    #[error("map section does not have {num_syms} definitions")]
+    InvalidMapSection { num_syms: usize },
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:647]: This should probably be InvalidMapSectionSize

Comment (alessandrod) [inline aya/src/obj/mod.rs:647]: This should probably be `InvalidMapSectionSize { size: usize }`

File: aya/src/obj/mod.rs [hunk 2]
```diff
@@ -494,6 +494,38 @@ impl Object {
...
 
+    fn parse_map_section(
+        &mut self,
+        section: &Section,
+        symbols: Vec<Symbol>,
+    ) -> Result<(), ParseError> {
+        if symbols.is_empty() {
+            return Err(ParseError::NoSymbolsInMapSection {});
+        }
+        let chunk_size = if section.size % symbols.len() as u64 == 0 {
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:505]: Can the `.maps` section contain variable length `bpf_map_def`s? (we need to check libbpf). See `MINIMUM_MAP_SIZE` which we support to support `iproute2` style maps.

File: aya/src/obj/mod.rs [hunk 3]
```diff
@@ -600,6 +642,15 @@ pub enum ParseError {
 
     #[error("map for section with index {index} not found")]
     MapNotFound { index: usize },
+
+    #[error("map section does not have {num_syms} definitions")]
+    InvalidMapSection { num_syms: usize },
+
+    #[error("no map name provided")]
+    NoMapName {},
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:650]:
```
#[error("the map number {i} in the `maps` section doesn't have a symbol name")]
MapSymbolNotFound { i: usize }
```

File: aya/src/obj/mod.rs [hunk 4]
```diff
@@ -600,6 +642,15 @@ pub enum ParseError {
...
+
+    #[error("no symbols in map section")]
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:652]: "no symbols found for the maps included in the `maps` section"

File: aya/src/obj/mod.rs [hunk 5]
```diff
@@ -517,9 +549,19 @@ impl Object {
...
+                    .values()
+                    .cloned()
+                    .filter(|s| {
+                        if let Some(idx) = s.section_index {
+                            idx == section.index
+                        } else {
+                            false
+                        }
+                    })
+                    .collect();
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:563]: I think this needs to be sorted by `s.address`?

File: aya/src/obj/mod.rs [hunk 6]
```diff
@@ -517,9 +549,19 @@ impl Object {
             BpfSectionKind::Btf => self.parse_btf(&section)?,
             BpfSectionKind::BtfExt => self.parse_btf_ext(&section)?,
             BpfSectionKind::Maps => {
-                let name = section.name.splitn(2, '/').last().unwrap();
-                self.maps
-                    .insert(name.to_string(), parse_map(&section, name)?);
+                let symbols = self
+                    .symbols_by_index
+                    .values()
+                    .cloned()
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:555]: Not that it makes much of a difference but cloned should probably be after filter

File: aya/src/obj/mod.rs [hunk 7]
```diff
@@ -1107,6 +1174,36 @@ mod tests {
...
+            map_type: 1,
+            key_size: 2,
+            value_size: 3,
+            max_entries: 4,
+            map_flags: 5,
+            ..Default::default()
+        })
+        .to_vec();
+        let mut buf = vec![];
+        buf.extend(map_data.iter());
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:1194]: &map_data instead of iter should work and be more readable

**Review by alessandrod — CHANGES_REQUESTED** [review]

File: aya/src/obj/mod.rs (hunk 2 above)
Comment (dave-tucker) [reply aya/src/obj/mod.rs:505]: `maps` can contain multiple map defs.... but they all MUST be the same size.
Here's the map counting/iteration logic and you can see the assumption in the comments.
https://github.com/libbpf/libbpf/blob/master/src/libbpf.c#L1968

File: aya/src/obj/mod.rs (hunk 5 above)
Comment (dave-tucker) [reply aya/src/obj/mod.rs:590]: Hmmm, why would we need to do that?

**Comment by dave-tucker** [comment]
@alessandrod ready for another look. This time around I decided to read into the section data from the address/size in the symbols table, so we can support definitions that are different lengths. The only drawback is that we don't verify all of the section data has been consumed...

File: aya/src/obj/mod.rs [hunk 8]
```diff
@@ -625,6 +665,15 @@ pub enum ParseError {
 
     #[error("map for section with index {index} not found")]
     MapNotFound { index: usize },
+
+    #[error("map section does not have {num_syms} definitions")]
+    InvalidMapSectionSize { num_syms: usize },
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:670]: This is unused now

File: aya/src/obj/mod.rs [hunk 9]
```diff
@@ -625,6 +665,15 @@ pub enum ParseError {
...
+
+    #[error("the map number {i} in the `maps` section doesn't have a symbol name")]
+    MapSymbolNotFound { i: usize },
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:673]: This should actually be `MapSymbolNameNotFound` - we have the symbol, just not a name for it.

File: aya/src/obj/mod.rs [hunk 10]
```diff
@@ -873,6 +922,22 @@ mod tests {
...
 
+    fn fake_sym(obj: &mut Object, section_index: usize, name: &str, size: u64) {
+        let idx = obj.symbols_by_index.len();
+        obj.symbols_by_index.insert(
+            idx + 1,
+            Symbol {
+                index: idx + 1,
+                section_index: Some(section_index),
+                name: Some(name.to_string()),
+                address: 0,
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:933]: The tests are parsing only the first map - you need to set the `address` field now

**Review by alessandrod — CHANGES_REQUESTED** [review]
couple of nits and fix tests pls

**Comment by dave-tucker** [comment]
@alessandrod nits addressed... and while I was editing the test I threw in some padding between the last `bpf_map_def`s just to check that the parsing logic would work as expected.

**Review by alessandrod — APPROVED** [review]

**Pull request merged**